### ISSUE TITLE
On improving the interface between "ImageDesc" and "ndarray".

### ISSUE THREAD

**Issue by KelSolaar** [AcademySoftwareFoundation/OpenColorIO#1206]
Hello,

As explained on [Slack](https://opencolorio.slack.com/archives/C3Q43TN4Q/p1605321436095300), I could not find an elegant way to have image data flowing between OCIO and Numpy.

At the moment to basically creating a `PackedImageDesc` not only requires passing the `ndarray` but also specifying the _width_ and _height_ and _channel_ count which seems superfluous because that data is already available directly from the shape of the array likewise, to get the data out, I need to reshape it. Here is a snippet illustrating the current code:

```python
IMAGE_DESC = ocio.PackedImageDesc(
    DIGITAL_LAD_ACESCG.astype(np.float32),
    WIDTH, HEIGHT, 3)

PROCESSOR.apply(IMAGE_DESC)

DIGITAL_LAD_ODT_sRGB = IMAGE_DESC.getData().reshape([HEIGHT, WIDTH, 3])
```

where `WIDTH` and `HEIGHT` are defined as follows:

```python
WIDTH, HEIGHT = DIGITAL_LAD_ACES2065_1.shape[1], DIGITAL_LAD_ACES2065_1.shape[0]
```

Ideally, for a Pythonista, the above snippet should be:

```python
IMAGE_DESC = ocio.PackedImageDesc(DIGITAL_LAD_ACESCG.astype(np.float32))

PROCESSOR.apply(IMAGE_DESC)

DIGITAL_LAD_ODT_sRGB = IMAGE_DESC.getData()
```

Or even better:

```python
PROCESSOR.apply(DIGITAL_LAD_ACESCG)
```

Changing the array in-place is mostly fine assuming warnings are explicit in the docs.

**Comment by michdolan** [comment]
The intent was that you could do this:

`processor.applyRGB(ndarray or list)` and `processor.applyRGBA(ndarray or list)`

And have that do in-place (returns reference too) modification of the array without an `ImageDesc`. That would be more in line with OCIO v1 behavior. I missed that goal on the first pass but the intent is to fix this by internally using an `ImageDesc` matching the ndarray and apply `processor.apply()`. @zachlewis noticed this as well and called it out in a doc meeting recently. 

Thanks for opening the issue. We'll definitely work on making this more intuitive.

**Comment by michdolan** [comment]
Also noting agreement on supporting passing ndarray directly to `processor.apply(ndarray)` without the `ImageDesc`

**Comment by KelSolaar** [comment]
Hello,

The first thing I tried was indeed `processor.applyRGB` but a) it did not visibly work on my image and b) the docs states it is for a single pixel.

Cheers,

Thomas

**Comment by zachlewis** [comment]
Indeed, I discovered the same thing with `processor.applyRGB` -- it seems to apply to only the first index of the array
(See #1185 ).

**Comment by remia** [comment]
Probably not something we want to support anyway but I noticed some arrays don't play nice with the current implementation. I was building a 3DLUT using `colour_science` library and the `lut.table` ndarray looks to not work out of the box. My current understanding is that it's related to the memory order as the channel order looks reversed.

```
lut = colour.LUT3D(size=64)

config = PyOpenColorIO.Config.CreateFromFile(config_path)
processor = config.getProcessor("ACES2065-1", "ACEScg").getDefaultCPUProcessor()
table = lut.table.astype(np.float32)
processor.applyRGB(table)
lut.table = table

colour.write_LUT(lut, "Rubbish.cube")
```

I saw some negative `ndarray.strides` for the last dimension of the colour LUT table so I suspect this may be a reason why the current implementation fails. Adding to `np.copy()` to force a positive stride solve the issue here while not being optimal solution obviously:
```
table = lut.table.astype(np.float32).copy()
```

Thought it was worth mentioning, I may dig deeper into the code at some point later.
